carmajava/signal_plugin/src/main/java/gov/dot/fhwa/saxton/carma/signal_plugin/ead/trajectorytree/ANAStarSolver.java

```diff
@@ -0,0 +1,270 @@
+/*
+ * Copyright (C) 2018 LEIDOS.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package gov.dot.fhwa.saxton.carma.signal_plugin.ead.trajectorytree;
+
+import gov.dot.fhwa.saxton.carma.signal_plugin.logger.ILogger;
+import gov.dot.fhwa.saxton.carma.signal_plugin.logger.LoggerManager;
+
+import java.util.*;
+
+import org.apache.commons.lang.mutable.MutableDouble;
+
+/**
+ * Implements the Anytime Non-parametric A* (ANA*) algorithm which can operate on nodes with distance, time, speed states
+ * The cost, heuristic, and goal is provided by the ICostModel
+ * The neighbors of a node is provided by the INeighborCalculator
+ * 
+ * ANA* algorithm is described http://goldberg.berkeley.edu/pubs/ana-aaai-2011-vdberg-shah-goldberg.pdf
+ */
+public class ANAStarSolver implements ITreeSolver {
+
+  protected static final ILogger log_ = LoggerManager.getLogger(AStarSolver.class);
+  protected long maxPlanningTimeMS = 500;
+  public static long iterationCount = 0;
+
+  @Override
+  public List<Node> solve(Node start, ICostModel costModel, INeighborCalculator neighborCalculator) {
+
+    // List of optimal parent from each node
+    // This map is used to extract the optimal path once the goal is found
+    final Map<Node,Node> cameFrom = new HashMap<>(1500);
+
+    // Each node's actual cost to reach from the start node
+    final Map<Node, Double> gScore = new HashMap<>(1500);
+
+    // Each node's heuristic cost to reach the goal
+    final Map<Node, Double> hScore = new HashMap<>(1500);
+
+    // Expected e-score cost to reach goal from start through each node
+    // eScore determines order of node expansion
+    // eScore = (G - gScore) / hScore
+    final Map<Node, Double> eScore = new HashMap<>(1500);
```

Comment:
Is this 1500 initial size based on observed data?

Reply:
Yes. I was seeing around 1000 elements in eScore so I went with 1500 as a buffer.